File: support/prototype/sheeter.go
package sheeter

import (
	"sync"
	"sync/atomic"
)

// NewSheeter 建立表格資料
func NewSheeter(loader Loader) *Sheeter {
	sheeter := &Sheeter{}
	sheeter.loader = loader
	sheeter.progress = NewProgress()
	return sheeter
}

// Sheeter 表格資料
type Sheeter struct {
	loader   Loader         // 裝載器物件
	progress *Progress      // 進度物件
	Alone0   HandmadeReader // 獨立表格說明
	Alone1   HandmadeReader // 獨立表格說明
	Merge0   HandmadeReader // 合併表格說明
	Merge1   HandmadeReader // 合併表格說明
}

// FromData 讀取資料處理
func (this *Sheeter) FromData() bool {
	this.progress.Reset()

	if this.loader == nil {
		return false
	} // if

	waitGroup := sync.WaitGroup{}
	waitGroup.Add(4)
	result := atomic.Bool{}
	result.Store(true)

	for _, itor := range []Reader{
		&this.Alone0,
		&this.Alone1,
	} {
		tmpl := itor

		go func() {
			defer waitGroup.Done()
			filename := tmpl.FileName()
			data := this.loader.Load(filename)

			if len(data) == 0 {
				return
			} // if

			if err := tmpl.FromData(data, true, this.progress); err != nil {
				this.loader.Error(filename.File(), err)
				result.Store(false)
			} // if
		}()
	} // for

	go func() {
		defer waitGroup.Done()

		for i, itor := range []Reader{
			&this.Alone0,
			&this.Alone1,
		} {
			filename := itor.FileName()
			data := this.loader.Load(filename)

			if len(data) == 0 {
				continue
			} // if

			if err := this.Merge0.FromData(data, i == 0, this.progress); err != nil {
				this.loader.Error(filename.File(), err)
				result.Store(false)
			} // if
		} // for
	}()

	go func() {
		defer waitGroup.Done()

		for i, itor := range []Reader{
			&this.Alone0,
			&this.Alone1,
		} {
			filename := itor.FileName()
			data := this.loader.Load(filename)

			if len(data) == 0 {
				continue
			} // if

			if err := this.Merge1.FromData(data, i == 0, this.progress); err != nil {
				this.loader.Error(filename.File(), err)
				result.Store(false)
			} // if
		} // for
	}()

	waitGroup.Wait()
	this.progress.Complete()
	return result.Load()
}

// Clear 清除資料
func (this *Sheeter) Clear() {
	this.progress.Reset()
	this.Alone0.Clear()
	this.Alone1.Clear()
	this.Merge0.Clear()
	this.Merge1.Clear()
}

// Progress 取得進度值
func (this *Sheeter) Progress() float32 {
	return this.progress.Get()
}

File: support/prototype/sheeter.cs
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Sheeter
{
    /// <summary>
    /// 表格資料
    /// </summary>
    public partial class Sheeter
    {
        public Sheeter(Loader loader)
        {
            this.loader = loader;
        }

        /// <summary>
        /// 讀取資料處理
        /// </summary>
        public async Task<bool> FromData()
        {
            progress.Reset();

            if (loader == null)
                return false;

            var task = new List<Task<bool>>();

            foreach (var itor in new Reader[] { this.Alone0, this.Alone1 })
            {
                var tmpl = itor;

                task.Add(
                    Task.Run(() =>
                    {
                        var filename = tmpl.FileName();
                        var data = loader.Load(filename);

                        if (data == null || data.Length == 0)
                            return true;

                        var error = tmpl.FromData(data, true, progress);

                        if (error.Length == 0)
                            return true;

                        loader.Error(filename.File, error);
                        return false;
                    })
                );
            } // for

            task.Add(
                Task.Run(() =>
                {
                    var first = true;
                    var result = true;

                    foreach (var itor in new Reader[] { this.Alone0, this.Alone1 })
                    {
                        var filename = itor.FileName();
                        var data = loader.Load(filename);

                        if (data == null || data.Length == 0)
                            continue;

                        var error = Merge0.FromData(data, first, progress);

                        if (error.Length != 0)
                        {
                            loader.Error(filename.File, error);
                            result = false;
                        } // if

                        first = false;
                    } // for

                    return result;
                })
            );
            task.Add(
                Task.Run(() =>
                {
                    var first = true;
                    var result = true;

                    foreach (var itor in new Reader[] { this.Alone0, this.Alone1 })
                    {
                        var filename = itor.FileName();
                        var data = loader.Load(filename);

                        if (data == null || data.Length == 0)
                            continue;

                        var error = Merge1.FromData(data, first, progress);

                        if (error.Length != 0)
                        {
                            loader.Error(filename.File, error);
                            result = false;
                        } // if

                        first = false;
                    } // for

                    return result;
                })
            );

            var result = await Task.WhenAll(task);

            progress.Complete();
            return result.All(itor => itor);
        }

        /// <summary>
        /// 清除資料
        /// </summary>
        public void Clear()
        {
            progress.Reset();
            Alone0.Clear();
            Alone1.Clear();
            Merge0.Clear();
            Merge1.Clear();
        }

        /// <summary>
        /// 取得進度值
        /// </summary>
        public float Progress()
        {
            return progress.Get();
        }

        /// <summary>
        /// 裝載器物件
        /// </summary>
        private readonly Loader loader;

        /// <summary>
        /// 進度物件
        /// </summary>
        private readonly Progress progress = new();

        /// <summary>
        /// 獨立表格說明
        /// </summary>
        public readonly HandmadeReader Alone0 = new();

        /// <summary>
        /// 獨立表格說明
        /// </summary>
        public readonly HandmadeReader Alone1 = new();

        /// <summary>
        /// 合併表格說明
        /// </summary>
        public readonly HandmadeReader Merge0 = new();

        /// <summary>
        /// 合併表格說明
        /// </summary>
        public readonly HandmadeReader Merge1 = new();
    }
}
